libzim creates (again) invalid title indexes

We have a new ifixit.com scraper which uses python-libzim. see [openzim/ifixit](https://github.com/openzim/ifixit)

A first test ZIM file has been created withthe Zimfarm and is available at https://mirror.download.kiwix.org/zim/.hidden/dev/ifixit_fr_all_2022-04.zim

But the ZIM file seems to have an invalid structure. Here is the `zimcheck` output:
```
[INFO] Checking zim file ifixit_fr_all_2022-04.zim
[INFO] Zimcheck version is 3.1.0
[INFO] Verifying ZIM-archive structure integrity...
Title index is not properly sorted.
Title index is not properly sorted.
  [ERROR] ZIM file's low level structure is invalid
```

Obviously this is a blocker!

--- a/include/zim/writer/creator.h
+++ b/include/zim/writer/creator.h
@@ -44,6 +44,11 @@ namespace zim
      *
      * During the creation of the zim file (and before the call to `finishZimCreation`),
      * some values must be set using the `set*` methods.
+     *
+     * All `add*` methods can throw a std::runtime_error exception if the entry
+     * cannot be added (mainly because a entry with the same path has already been added).
+     * It is up to the user to catch this exception and handle the error.
+     * The creator is still in a valid state and the creation can continue.
      */
     class Creator
     {

--- a/src/writer/creator.cpp
+++ b/src/writer/creator.cpp
@@ -498,19 +498,17 @@ namespace zim
           existing->markRemoved();
           dirents.insert(dirent);
         } else {
-          std::cerr << "Impossible to add " << NsAsChar(dirent->getNamespace()) << "/" << dirent->getPath() << std::endl;
-          std::cerr << "  dirent's title to add is : " << dirent->getTitle() << std::endl;
-          std::cerr << "  existing dirent's title is : " << existing->getTitle() << std::endl;
-          return;
+          std::ostringstream ss;
+          ss << "Impossible to add " << NsAsChar(dirent->getNamespace()) << "/" << dirent->getPath() << std::endl;
+          ss << "  dirent's title to add is : " << dirent->getTitle() << std::endl;
+          ss << "  existing dirent's title is : " << existing->getTitle() << std::endl;
+          throw std::runtime_error(ss.str());
         }
       };
 
-      // If this is a redirect, we're done: there's no blob to add.
-      if (dirent->isRedirect())
-      {
+      if (dirent->isRedirect()) {
         unresolvedRedirectDirents.insert(dirent);
         nbRedirectItems++;
-        return;
       }
     }
 

--- a/src/writer/creator.cpp
+++ b/src/writer/creator.cpp
@@ -495,6 +495,7 @@ namespace zim
         if (existing->isRedirect() && !dirent->isRedirect()) {
           unresolvedRedirectDirents.erase(existing);
           dirents.erase(ret.first);
+          existing->markRemoved();
           dirents.insert(dirent);
         } else {
           std::cerr << "Impossible to add " << NsAsChar(dirent->getNamespace()) << "/" << dirent->getPath() << std::endl;

--- a/test/creator.cpp
+++ b/test/creator.cpp
@@ -180,7 +180,8 @@ TEST(ZimCreator, createZim)
   creator.startZimCreation(tempPath);
   creator.addRedirection("foo", "WrongRedirection", "foobar", {{zim::writer::FRONT_ARTICLE, true}}); // Will be replaced by item
   auto item = std::make_shared<TestItem>("foo", "Foo", "FooContent");
-  creator.addItem(item);
+  EXPECT_NO_THROW(creator.addItem(item));
+  EXPECT_THROW(creator.addItem(item), std::runtime_error);
   // Be sure that title order is not the same that url order
   item = std::make_shared<TestItem>("foo2", "AFoo", "Foo2Content");
   creator.addItem(item);

--- a/test/creator.cpp
+++ b/test/creator.cpp
@@ -178,6 +178,7 @@ TEST(ZimCreator, createZim)
   creator.setUuid(uuid);
   creator.configIndexing(true, "eng");
   creator.startZimCreation(tempPath);
+  creator.addRedirection("foo", "WrongRedirection", "foobar", {{zim::writer::FRONT_ARTICLE, true}}); // Will be replaced by item
   auto item = std::make_shared<TestItem>("foo", "Foo", "FooContent");
   creator.addItem(item);
   // Be sure that title order is not the same that url order
@@ -188,7 +189,7 @@ TEST(ZimCreator, createZim)
   creator.addIllustration(96, "PNGBinaryContent96");
   creator.setMainPath("foo");
   creator.addRedirection("foo3", "FooRedirection", "foo"); // No a front article.
-  creator.addRedirection("foo4", "FooRedirection", "NoExistant"); // Invalid redirection, must be removed by creator
+  creator.addRedirection("foo4", "FooRedirection", "NoExistant", {{zim::writer::FRONT_ARTICLE, true}}); // Invalid redirection, must be removed by creator
   creator.finishZimCreation();
 
   // Do not use the high level Archive to test that zim file is correctly created but lower structure.

